numba/compiler_machinery.py

```diff
@@ -0,0 +1,376 @@
+        """
+        return False
+
+    def get_analysis_usage(self, AU):
+        """Override to set analysis usage
+        """
+        pass
+
+    def get_analysis(self, pass_name):
+        return self._analysis[pass_name]
+
+
+class FunctionPass(CompilerPass):
+    pass
+
+
+class AnalysisPass(CompilerPass):
+    pass
+
+
+class LoweringPass(CompilerPass):
+    pass
+
+
+class AnalysisUsage:
+    """This looks and behaves like LLVM's AnalysisUsage because its like that.
+    """
+
+    def __init__(self):
+        self._required = set()
+        self._preserved = set()
+
+    def get_required_set(self):
+        return self._required
+
+    def get_preserved_set(self):
+        return self._preserved
+
+    def add_required(self, pss):
+        self._required.add(pss)
+
+    def add_preserved(self, pss):
+        self._preserved.add(pss)
+
+    def __str__(self):
+        return "required: %s\n" % self._required
+
+
+_DEBUG = False
+
+
+def debug_print(*args, **kwargs):
+    if _DEBUG:
+        print(*args, **kwargs)
+
+
+pass_timings = namedtuple('pass_timings', 'init run finalise')
+
+
+class PassManager(object):
+    """
+    The PassManager is a named instance of a particular compilation pipeline
+    """
+    # TODO: Eventually enable this, it enforces self consistency after each pass
+    _ENFORCING = False
+
+    def __init__(self, pipeline_name):
+        """
+        Create a new pipeline with name "pipeline_name"
+        """
+        self.passes = []
+        self.exec_times = OrderedDict()
+        self._finalized = False
+        self._analysis = None
+        self._print_after = None
+        self.pipeline_name = pipeline_name
+
+    def _validate_pass(self, pass_cls):
+        if not (isinstance(pass_cls, str) or (inspect.isclass(pass_cls) and issubclass(pass_cls, CompilerPass))):
+            msg = "Pass must be referenced by name or be a subclass of a CompilerPass. Have %s" % pass_cls
+            raise TypeError(msg)
+        if isinstance(pass_cls, str):
+            pass_cls = _pass_registry.find_by_name(pass_cls)
+        else:
+            if not _pass_registry.is_registered(pass_cls):
+                raise ValueError("Pass %s is not registered" % pass_cls)
+
+    def add_pass(self, pss, description=""):
+        """
+        Add a pass to the PassManager
+        """
+        self._validate_pass(pss)
+        func_desc_tuple = (pss, description)
+        self.passes.append(func_desc_tuple)
+        self._finalized = False
+
+    def add_pass_after(self, pass_cls, location):
+        """
+        Add a pass to the PassManager after the pass "location"
+        """
+        assert self.passes
+        self._validate_pass(pass_cls)
+        self._validate_pass(location)
+        for idx, (x, _) in enumerate(self.passes):
+            if x == location:
+                break
+        else:
+            raise ValueError("Could not find pass %s" % location)
+        self.passes.insert(idx + 1, (pass_cls, str(pass_cls)))
```

Comment:
It could be, I was maintaining the original interface style, is a description useful in this new style as passes are uniquely defined and have a name? Am happy to do either.